opendap_protocol/protocol.py

```diff
@@ -496,17 +495,21 @@ def dods_encode(data, dtype):
 
     yield packed_length
 
+    chunk_size = int(Config.STREAMING_BLOCK_SIZE / data.dtype.itemsize)
     if isinstance(data, da.Array):
-        # Encode in chunks of a defined size if we work with dask.Array
-        chunk_size = int(Config.DASK_ENCODE_CHUNK_SIZE / data.dtype.itemsize)
-        flat = data.ravel()
-        for start in range(0, data.size, chunk_size):
-            block = flat[slice(start, chunk_size)]
-            yield block.astype(dtype.str).compute().tobytes()
-    else:
-        # Make sure we always encode an array or we will get wrong results
-        data = np.asarray(data)
-        yield data.astype(dtype.str).tobytes()
+        data = data.ravel()
+
+    for start in range(0, data.size, chunk_size):
+        end = start + chunk_size
+        if isinstance(data, da.Array):
+            block = data[slice(start, end)].compute()
+        elif has_xarray and isinstance(data, Variable):
+            npidxr = np.unravel_index(np.arange(start, min(end, data.size)), shape=data.shape)
```

Comment:
<img width="700" alt="image" src="https://github.com/user-attachments/assets/46c6b8c4-fefe-4021-91a4-75ca4bbd558c">
for 30MB blocks,